synapse/handlers/pagination.py

```diff
@@ -176,6 +136,82 @@ def __init__(self, hs: "HomeServer"):
                     job.longest_max_lifetime,
                 )
 
+        if self._is_master:
+            self.clock.looping_call(
+                run_as_background_process,
+                PaginationHandler.PURGE_ROOMS_PERIOD,
+                "purge_rooms",
+                self.purge_rooms,
+            )
```

Comment:
What happens if `purge_rooms(...)` takes longer to run than the period? Is `looping_call(...)` smart enough to not start the next call until the first finishes?

We skip a shutdown/purge for a room if it already exists in` self._delete_by_id` (meaning we're already working on it). Probably still a chance to race because of async things.

But we still could start a lot more purges in parallel. Imagine `purge_rooms(...)` takes 15 hours and each hour we start another `purge_rooms(...)` which will pick up the next room and by the end we would have 15 purges going on when we only expect 1 to be going on at a time.